espaloma/data/dataset.py

```diff
@@ -0,0 +1,186 @@
+        if self.graphs is None:
+            raise RuntimeError('Empty molecule dataset.')
+
+        if isinstance(idx, int): # sinlge element
+            if self.transforms is None: # when no transform act like list
+                return self.graphs[idx]
+
+            else:
+                graph = self.graphs[idx]
+
+                # nested transforms
+                for transform in self.transforms:
+                    graph = transform(graph)
+
+                return graph
+
+        elif isinstance(idx, slice): # implement slicing
+            if self.transforms is None:
+                # return a Dataset object rather than list
+                return self.__class__(graphs=self.graphs[idx]) 
+            else:
+                graphs = []
+                for graph in self.graphs:
+
+                    # nested transforms
+                    for transform in transforms:
+                        graph = transform(graph)
+                    graphs.append(graph)
+
+                return self.__class__(graphs=graphs)
+
+    def __iter__(self):
+        if self.transforms is None:
+            return iter(self.graphs)
+
+        else:
+            # TODO:
+            # is this efficient?
+            graphs = iter(self.graphs)
+            for transform in self.transforms:
+                graphs = map(
+                        transform,
+                        graphs)
+
+            return graphs
+
+    def apply(self, fn, in_place=False):
+        assert callable(fn)
+        assert isinstance(in_place, bool)
+
+        if in_place is False: # add to list of transforms
+            if self.transforms is None:
+                self.transforms = []
+
+            self.transforms.append(fn)
+
+        else: # modify in-place
+            self.graphs = list(map(fn, self.graphs))
+
+        return self # to allow grammar: ds = ds.apply(...)
+
+    def split(self, partition):
+        """ Split the dataset according to some partition.
+
+        """
+        n_data = len(self)
+        partition = [int(n_data * x / sum(partition)) for x in partition]
+        ds = []
+        idx = 0
+        for p_size in partition:
+            ds.append(self[idx : idx + p_size])
+            idx += p_size
+
+        return ds
+
+class GraphDataset(Dataset):
+    """ Dataset with additional support for only viewing
+    certain attributes as `torch.utils.data.DataLoader`
+
+    
+    """
+
+    def __init__(self, graphs, first=None):
+        super(GraphDataset, self).__init__()
+        from openforcefield.topology import Molecule
+
+        if all(
+                isinstance(
+                    graph, 
+                    Molecule
+                ) or isinstance(
+                    graph,
+                    str) for graph in graphs):
+
+            if first is None:
+                graphs = [esp.Graph(graph) for graph in graphs]
+            else:
+                graphs = [esp.Graph(graph) for graph in graphs[:first]]
```

Comment:
document intent of `first` argument and `[:first]` slice -- for example, why not call constructor with different input graphs, `GraphDataset(graphs)` vs `GraphDataset(graphs[:first])`

Reply:
ah yeah this looks a bit redundant indeed

Reply:
okay I think I was thinking about adding it to the pre-written collection of datasets